Formatted output for cluster options (part 1)

nrwahl2 opened the pull request:
This part can stand on its own before we continue moving toward proper `pcmk__output_t` formatted output.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -939,68 +939,80 @@ pcmk__cluster_option(GHashTable *options, const char *name)
-        if (!pcmk__str_empty(escaped_en)
-            && (s->str[s->len - 1] != '.')
-            && ((s->str[s->len - 2] != '.') || (s->str[s->len - 1] != ')'))) {
+        // @TODO Use XML node for allowed values
```
On second thought, the `<option>` elements that get created for `"select"`-type options make the `"Allowed values: "` string section redundant. We could probably drop it and make this `add_desc()` function much simpler.

There's at least one non-`select` option (`cluster-recheck-interval`) that has an allowed values string. It's not a list of enumerated values. For any options like that, I can just copy the allowed values string into the longdesc. That's more appropriate.

nrwahl2 replied:
Done

clumens commented on cts/cli/regression.daemons.exp:
```diff
@@ -1,435 +1,747 @@
 =#=#=#= Begin test: Get CIB manager metadata =#=#=#=
-<?xml version=""?>
```
This is required for XML 1.1 documents (https://www.w3.org/TR/xml11/#sec-prolog-dtd) but apparently not for older versions.  I don't know which we are using offhand, but I'd lean towards making sure it's included just to be clear.

nrwahl2 replied:
Our API output already doesn't include it though.
```
reid@laptop:~/git/pacemaker$ sudo crm_mon --output-as=xml
<pacemaker-result api-version="2.32" request="crm_mon --output-as=xml">
  <summary>
...
```

The goal is to be using an XML `pcmk__output_t` object by the time this PR is finished; this is just an intermediate step.

kgaillot replied:
The output object should probably display it, but that doesn't need to be part of this

nrwahl2 replied:
I'm fine with that, but yeah if we display it then it needs to be done at the output object level.

nrwahl2 replied:
https://projects.clusterlabs.org/T767

nrwahl2 replied:
> This is required for XML 1.1 documents (https://www.w3.org/TR/xml11/#sec-prolog-dtd) but apparently not for older versions. I don't know which we are using offhand, but I'd lean towards making sure it's included just to be clear.

For an unrelated reason, I just got confirmation that libxml only supports XML 1.0.
https://discourse.gnome.org/t/does-xmlnewdoc-support-xml-1-1/19329

clumens commented on lib/common/options.c:
```diff
@@ -939,68 +939,80 @@ pcmk__cluster_option(GHashTable *options, const char *name)
         static const char *locale = NULL;
+        const char *local_desc = _(desc);
 
-        char *localized = crm_xml_escape(_(desc));
```
I assume this is why the regression test output is no longer escaped.  I'd also keep this since who knows what changing it will break.

Incidentally, there's a comment in crm_xml_escape from 2013 about how xmlNodeDump is slow.  I wonder if that's still true (or was ever true) and if we therefore even need to keep doing our own escaping instead of asking libxml to do it for us.

nrwahl2 replied:
> I'd also keep this since who knows what changing it will break.

In general, we don't escape the contents of XML text nodes. See `pcmk__xml2text()` -> `dump_xml_text()`. Our high-level dump functions all call `pcmk__xml2text()`. That's what makes me feel a lot more comfortable about this.

There are a few exceptions. We're not really consistent, so it's not clear to me why we escape the output in some certain cases. Sometimes we do our own escaping, sometimes not. ometimes we build a string manually, sometimes we tell libxml to dump it for us (which caused some buffering issues; see next major point below). I'm not going to blindly assume that we HAVE a good reason for everything. We do escape the following:
* XML attribute values -- see `pcmk__dump_xml_attr()` and `show_xml_element()`
* the exit reason in the XML version of the `"failed-action"` message
  * This becomes an XML attribute value. Until c63a898b, it would have been escaped by the "XML attribute values" case, so this would've been entirely redundant. However, with that commit, we started using `pcmk__xml2fd()` -> `xmlNodeDumpOutput()` in `lib/common/output_xml.c`, like poki wanted to do. The libxml dump functions don't escape all the special characters that `crm_xml_escape()` does.
* the description of a systemd unit in its text OCF metadata (see `systemd_unit_metadata()`)

Everything else goes unescaped.

> Incidentally, there's a comment in crm_xml_escape from 2013 about how xmlNodeDump is slow. I wonder if that's still true (or was ever true) and if we therefore even need to keep doing our own escaping instead of asking libxml to do it for us.

Yeah... that comment saying `xmlNodeDump()` is too slow is from beekhof. Then poki drafted c63a898b as mentioned above, which started using `xmlNodeDumpOutput()` (the modern version of `xmlNodeDump()`). His argument was that using `xmlNodeDumpOutput()` would make things faster. So who even knows?

1382915a (from you) reverts most of poki's commit in order to fix a regression that it caused. The only place where we still use `pcmk__xml2fd()` is when finishing XML formatted output.

nrwahl2 replied:
@kgaillot tag as an FYI on these inconsistencies, even though I don't think they're relevant here since this is consistent with our current practices for text nodes. You might already be aware of all this.

kgaillot replied:
I'm pretty sure it does need to be escaped. If someone puts a `<` in the description, it will break the XML if not escaped. If we're not escaping text nodes elsewhere, either that's a bug or we know the content doesn't need escaping.

nrwahl2 replied:
> I'm pretty sure it does need to be escaped. If someone puts a `<` in the description, it will break the XML if not escaped.

Confirmed. Pacemaker fails to build if you put that into one of the options. I added `<test>` to the content of a longdesc and `make` failed with
```
/home/reid/git/pacemaker/daemons/controld/pacemaker-controld metadata > pacemaker-controld.xml
/usr/bin/xsltproc --nonet --novalid --stringparam man.name pacemaker-controld \
	 ../../xml/ocf-meta2man.xsl \
	/home/reid/git/pacemaker/daemons/controld/pacemaker-controld.xml > /home/reid/git/pacemaker/daemons/controld/pacemaker-controld.dbook
/home/reid/git/pacemaker/daemons/controld/pacemaker-controld.xml:18: parser error : Opening and ending tag mismatch: test line 17 and shortdesc
      </shortdesc>
...
```

> If we're not escaping text nodes elsewhere, either that's a bug or we know the content doesn't need escaping.

I'd say it's a bug then.
* `dump_xml_text()` does our general-purpose XML-to-text dumping for text nodes, and it doesn't escape text node contents. We could say we're assuming the text got escaped when we added it to the node. However, we escape XML attribute values in `pcmk__dump_xml_attr()` -- so we don't assume it for attribute values.
* `show_xml_node()`/`show_xml_element()` (the formatted output display functions) appear to skip text nodes entirely.
* systemd metadata *does* get `crm_xml_escape()`'d.
* LSB and st_lha metadata... might? They call `xmlEncodeEntitiesReentrant()`... which at a glance looks like a libxml-provided equivalent of `crm_xml_escape()`. That's probably worth looking into further to compare and contrast.

I have no idea about the state of `xmlNodeDumpOutput()` as discussed above... conflicting statements, obfuscated comments and commit messages, and a history of unexpected behavior (see the buffering issue).

nrwahl2 replied:
https://projects.clusterlabs.org/T768

nrwahl2 commented:
retest this please

nrwahl2 commented:
Fix Coverity error (unfreed var)

nrwahl2 commented:
Trying to leave this PR alone, pending review. Ongoing work is in https://github.com/nrwahl2/pacemaker/tree/nrwahl2-T620_new.

kgaillot commented on lib/common/options.c:
```diff
@@ -1051,48 +1064,57 @@ pcmk__format_option_metadata(const char *name, const char *desc_short,
-        add_desc(s, PCMK_XE_LONGDESC, opt_desc_long, "      ");
-        add_desc(s, PCMK_XE_SHORTDESC, opt_desc_short, "      ");
+        if (add_desc(parameter, PCMK_XE_LONGDESC,
+                     opt_desc_long) != pcmk_rc_ok) {
```
You checked NULL before doing add_desc() above ... if you do it in add_desc(), we don't have to worry about it

nrwahl2 replied:
That's because the standard doesn't require longdesc and shortdesc at the top-level. They're required only for parameters.

So at the top-level, we add the description only if the description exists. At the parameter level, we assert that at least one of the descriptions is non-NULL. If only one is non-NULL, we use it for both.

FYI, in 72a254a8 (#3338), I assert non-NULL at the top level too, even though the standard doesn't require it. It's easier that way. We're in control of the metadata and we can guarantee that the longdesc and shortdesc are both present at the top level.

kgaillot replied:
yep that's fine

kgaillot commented on lib/common/options.c:
```diff
@@ -987,6 +987,85 @@ add_desc(xmlNode *parent, const char *tag, const char *desc)
+    }
+
+    // The standard requires a parameter type
+    CRM_ASSERT(option->type != NULL);
```
Since we're returning and checking rich error codes, just return EINVAL instead of the two assertions

nrwahl2 replied:
It's a programming error since we're in control of the metadata, so I don't think it would hurt to catch it via an assert... though I can see an argument for not wanting to assert within a daemon, even if it is a soon-to-be-deprecated metadata call.

FWIW, in an upcoming commit (72a254a8), we no longer return rich error codes. It doesn't make sense to NULL-check created XML nodes and return `ENOMEM`, since the XML output functions would've segfaulted by then. So most of the error checking is dropped as of right now.

nrwahl2 commented:
Added two commits to the beginning that address the escaping issues, and then updates tests, etc.

---

Our escaping is more aggressive than it needs to be. In particular, we don't need to escape tabs and newlines, and we need to escape single and double quotes only for attribute values (not for text node content). Technically, even for attribute values we only need to escape the one we're using to delimit the value -- for example:
```
attr="va'lue" // OK
attr="va\"lue" // OK
attr="va"lue" // bad
attr='va"lue' // OK
attr='va\'lue' // OK
attr='va'lue' // bad
```

We also don't need to allocate a new string unless the base string contains special characters. We could use something like `strcspn()` to determine if a string needs to be escaped.

After all of this is done, I'll probably make an internal version of `crm_xml_escape()` that's only as aggressive as it needs to be, and deprecate `crm_xml_escape()`.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -669,13 +669,36 @@ create_xml_node(xmlNode * parent, const char *name)
+ *
+ * \note \c xmlNodeSetContent() does not escape special characters.
+ */
+void
```
This could return a value since `xmlNodeSetContent()` returns `1` for generic error and `-1` for memory allocation failure. I kept it simple for now since the existing code doesn't do error-checking, but I could add it.

kgaillot commented:
> After all of this is done, I'll probably make an internal version of `crm_xml_escape()` that's only as aggressive as it needs to be, and deprecate `crm_xml_escape()`.

sounds good

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -669,13 +669,36 @@ create_xml_node(xmlNode * parent, const char *name)
+        free(escaped);
+    } else {
+        xmlNodeSetContent(node, (pcmkXmlStr) content);
+    }
```
Ended up checking character-by-character because we also need to check ranges.

nrwahl2 commented:
Updated to address review and replace crm_xml_escape(). About to rebase on main to resolve conflict.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1350,16 +1367,17 @@ crm_xml_escape(const char *text)
     copy = strdup(text);
     CRM_ASSERT(copy != NULL);
-    for (size_t index = 0; index < length; index++) {
+    for (size_t index = 0; index <= length; index++) {
```
This should still be `< length`. The test was off by one in the first place. The new `pcmk__xml_escape()` uses `< length`.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1333,6 +1333,108 @@ replace_text(char *text, int index, size_t *length, const char *replace)
+                    /* Escape non-printing and Unicode characters (0x7f (delete)
+                     * is ASCII but non-printing)
+                     *
+                     * @TODO Handle multi-byte Unicode characters
```
I really just don't want to deal with this right now. This preserves existing behavior. But since this we're working with a signed char, the value will never be higher than `0x7f`.

kgaillot replied:
Yep that's fine

kgaillot commented on lib/common/xml.c:
```diff
@@ -1333,6 +1333,108 @@ replace_text(char *text, int index, size_t *length, const char *replace)
+ *       https://discourse.gnome.org/t/intended-use-of-xmlencodeentitiesreentrant-vs-xmlencodespecialchars/19252
+ */
+char *
+pcmk__xml_escape(const char *text, bool escape_quote)
```
worth a unit test

nrwahl2 replied:
I wish there already were one for `crm_xml_escape()` to use as a template... but you're right

kgaillot commented on daemons/based/pacemaker-based.c:
```diff
@@ -134,11 +134,11 @@ based_metadata(void)
+    char *s = pcmk__cluster_option_metadata(name, desc_short, desc_long,
+                                            pcmk__opt_context_based);
 
     printf("%s", s);
```
All of the daemons should be using output_xml() (like pacemaker-schedulerd) instead of printf()

nrwahl2 replied:
That's coming, I just postponed it to later.

https://github.com/ClusterLabs/pacemaker/pull/3338/commits/1a8c6e2f

kgaillot replied:
Ah cool. However we don't need to create an output object, all the daemons already have one for their --version argument etc. We don't want to force XML output; it's a bit confusing, but any output format can display XML via output_xml(): text outputs it raw (which is important for pcs) while xml wraps it in CDATA so it doesn't break the API schema.

nrwahl2 replied:
It's done that way for backward compatibility. Currently, a daemon outputs raw XML for "text" output and CDATA for "xml" output.

As of https://github.com/ClusterLabs/pacemaker/commit/1a8c6e2f, `pcmk__cluster_option_metadata()` accepts an "xml" output object and uses it to output raw XML. We want the daemon output to stay the same (for example if the daemon uses "text" output). So we create a temporary XML output object that doesn't print anything, make a copy of its result, and call `output_xml()` on the result using the daemon's main output object. It's inefficient, but these are one-shot calls.

Later, we turn `pcmk__cluster_option_metadata()` into an output message function and then add a default one for "text": https://github.com/ClusterLabs/pacemaker/pull/3338/commits/e363cee1. The "text" one is a human-readable list, as we discussed previously.

nrwahl2 replied:
Huh. What I said above is MOSTLY correct. However, I thought in that future commit, we called `output_xml()` on the resulting `PCMK_XE_RESOURCE_AGENT` XML object. Instead we still call `print()` for the controller, CIB manager, and fencer.

Technically that's still the most backward-compatible way to do it, since those two daemons have always printed raw XML to stdout even if the main output object is xml-format... it would probably be fine to call `output_xml` on it though (and wouldn't be any more complex).

kgaillot replied:
I think I get it now ... for the new command, we want the text output to be friendly, so that's what your custom message will do (not output_xml()). For the daemons to preserve their behavior, they'll need to call output_xml() directly instead of the custom message. That will preserve the current behavior for text and none, and output XML in an API-standard way. (There's no schema for daemon XML output, so we don't have to worry about that.)

nrwahl2 replied:
Right :)

kgaillot commented on lib/common/xml_attr.c:
```diff
@@ -78,13 +78,15 @@ pcmk__dump_xml_attr(const xmlAttr *attr, GString *buffer)
     value = (const char *) attr->children->content;
+    if (value == NULL) {
+        return;
+    }
```
An empty value (`<e a=""/>`) would be better in case the XML has a schema that requires the attribute

nrwahl2 replied:
That's a reasonable concern, but an empty string is a value isn't it? The flip side of "there may be a schema that requires this" is:

"There may be a caller that treats an absent attribute differently from an attribute whose value is the empty string, and the relevant schema doesn't require the attribute."

kgaillot replied:
Hmm, less than ideal either way. I suppose it doesn't matter which we pick. Maybe add a comment describing the choice.

kgaillot reviewed (commented):
looking good

nrwahl2 commented:
Rebased on current main; no code changes

nrwahl2 commented:
Addressed review:
* Inline-functionize the `is_unicode()` check
* Unit tests for `pcmk__xml_escape()` and `pcmk__xml_needs_escape()`
* Fix issue found by unit test in `pcmk__xml_needs_escape()` (not a huge one, but it wanted to escape newlines and tabs)
* Add comment to `pcmk__dump_xml_attr()`

kgaillot commented on lib/common/xml.c:
```diff
@@ -1315,6 +1315,23 @@ write_xml_file(const xmlNode *xml, const char *filename, gboolean compress)
 
+/*!
+ * \internal
+ * \brief Check whether a string's first two bytes represent a Unicode charcter
```
Technically we're checking for multibyte UTF-8 characters

kgaillot commented on lib/common/xml.c:
```diff
@@ -1315,6 +1315,23 @@ write_xml_file(const xmlNode *xml, const char *filename, gboolean compress)
+static inline bool
+is_unicode(const char *text)
+{
+    // @TODO Is this a valid and complete test for multi-byte Unicode chars?
```
Mostly: https://stackoverflow.com/questions/5012803/test-if-char-string-contains-multibyte-characters

I think the only problem is assuming there will be an even number of bytes. It can be 2, 3, or 4.

kgaillot commented on lib/common/xml.c:
```diff
@@ -1463,9 +1484,11 @@ pcmk__xml_escape(const char *text, bool escape_quote)
+            /* Skip two bytes (one here, one in loop counter).
+             *
+             * @TODO Better to escape these? But readability concerns for
+             * languages with non-ASCII alphabets.
```
There shouldn't be any reason to escape them, as long as we're always using UTF-8 encoding, which I believe we are, though we probably could do more to advertise it correctly, in particular using the XML prolog (`<?xml version = "1.0" encoding = "UTF-8" standalone = "no" ?>`)

nrwahl2 replied:
> no reason

Nice. We should be *able* to escape them, but escaping them (albeit incorrectly) was why that upstream contributor added the "skip" block in the first place. It was making Chinese text unreadable.

If it's valid to use UTF-8 characters, I'd prefer to do so.

> prolog

Added to description of T767.

kgaillot reviewed (commented):
Unit tests rock :)

kgaillot reviewed (approved):
awesome

nrwahl2 commented:
```
+  * Restart    rabbitmq:0                 (        rabbitmq-bundle-0 )  due to resource definition change
+  * Restart    rabbitmq:1                 (        rabbitmq-bundle-1 )  due to resource definition change
+  * Restart    rabbitmq:2                 (        rabbitmq-bundle-2 )  due to resource definition change
```
et cetera

are. you. kidding. me.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1317,19 +1317,45 @@ write_xml_file(const xmlNode *xml, const char *filename, gboolean compress)
+     *   '0' bit.
+     * * In multi-byte UTF-8 characters, each byte begins with a '1' bit.
+     */
+    for (uint8_t *byte_8 = (uint8_t *) text; (*byte_8 & 0x80) != 0; byte_8++) {
```
I revamped this and am about to push a new version. I realized that in rare cases where `CHAR_BIT != 0` and we find one UTF-8 character, the next character we check might not be at a C byte (char) boundary.

For example, if we have a 24-bit UTF-8 character and `CHAR_BIT` is 16, then we probably need to start the next check at bit 32.

nrwahl2 commented:
retest this please

nrwahl2 commented:
Coverity is throwing a dead code error because I have
```
        if ((utf8_bits % CHAR_BIT) > 0) {
            c_bytes++;
        }
```

CPP is expanding `CHAR_BIT` to 8 before Coverity analyzes it. So, I'll have to guard it. Stupid...
